### ISSUE TITLE
Separate snapshots storage, engine handling and topologies

### ISSUE THREAD

**Issue by jhprinz** [choderalab/openpathsampling#208]
@dwhswenson and me discussed that it would be nice if we could _outsource_ the storage of snapshots. This is particular necessary when we want to use other engines, see #144 and be compatible with our current storage system.

#### basic idea
All of openpathsampling is completely independent of what the snapshots represent as long as the engine knows how to generate these given a starting snapshot and we know how to create collective variables for it.
Right now we already have two types of systems : OpenMM (MD) and ToyDynamics, which are quite different, but only on the snapshot side. And the question is: Do we try to make a very general snapshot class that can handle all cases, e.g.
- Additional metainformation (QM states, etc...)
- w/o velocities (might be brownian dynamics)
- w/o kinetic/potential energy
- an external engine runs the dynamics and already stores the frames somewhere

or do we leave that to the `engine` to know about the specifics of a snapshot?
Right now the code is independent of the snapshot. Only the `snapshot_store.py`, and `snapshot.py` know about it. Even collective variables are made specific for a certain type of snapshot.

#### specific questions
I think there are a few questions that we should decide before doing this (if we want to). 

1. Which object knows about the structure of a `Snapshot` or where in the code do we put/hide this information. Most obvious is the `engine` class, which would fit perfectly to the external engine class in #144. It knows specifics about the simulation (e.g. topology, integrators, number of atoms, etc...) and it knows how the snapshots are stored (e.g. numpy array in memory, frams in gromacs trajectory file, etc...)
This would look like
    ```
    # setup our engine with all we now
    engine = GromacsEngine(options)
    engine.moreoptions = ...
    # tell the ops_storage how snapshots are handled
    storage = Storage('filename.nc', engine.snapshot_store) 
    # we can still store the engine with all its settings in the storage
    storage.save(engine)
    ```
    The other possibility would be to have a separate `SystemDescription` object that knows this information, but it would be again engine specific, so I prefer that the engine is the main player here.

2. Do we allow to create `Snapshot`s on our own or only via the engine?
In the examples we often create the first snapshot manually. This would not be
a problem, but if you want to store these we would need to implement storage of frames independent of what the engine generates in files. Just saying that we could avoid this.

3. If we allow creating new `Snapshot`s, do we want to store them with the newly generated trajectories or store these in the main ops_store?
It could make saving a little more clean instead of creating initial single frame trajectories every time we
start an engine with a new trajectory. These potential new snapshots would be saved in the ops_storage instead, while the engine creates the initial files for the simulation, runs and afterwards only the trajectory file is kept. All other files are considered temporary and will be deleted. This would also allow to have a fall back to store also Gromacs, Lammps, etc the way we do this now or mix. The downside is that we need to implement up to 2 stores for snapshots, one that knows how to store in the ops netcdf file and one that knows how to handle the external storage used by the engine. This can of course vary by case:
    - OpenMM might offer internal storage or use of hdf5 / mdtraj separate files,
    - ToyDynamics might use only internal
    - Gromacs also offers both / mixed support.
    ```
    engine = GromacsEngine()
    storage = Storage('file.nc', engine.internal_snapshot_store)
    ```
    In most cases we will only have one manual Snapshot (probably read from an external file) and all other snapshots are created by the engine.

#### when do we want to implement this
Considering that David has this engine stuff almost ready, I am tempted to say I can implement the separate storage in a few days and we would have this ready for 1.0. This could be very appealing for a lot of people even if we cannot run in parallel yet. From my point this is actually rather simple to implement and would be very powerful.

#### what we would need to do
- [ ] change initialization of Storage to require a snapshot_store and not a template. This actually makes the storage simpler since the logic to handle the topology, number of atoms, etc... if now part of the engine and engine can simply be stored as any other object
- [ ] write a snapshot_store that uses the external engine API
- [ ] write a snapshot_store for toy_dynamics and OpenMMEngine that uses the ops_store as before (this is practically copy and paste)
- [ ] maybe separate engine stuff in submodules (see #197) so each engine has its internal logic (integrators, etc...) an optional topology (like for OpenMM, but could be replaced by the OpenMM internal pickling), some utiliy functions and (this is new) a store_logic

**Comment by jhprinz** [comment]
Part of PR #197 .

**Comment by jhprinz** [comment]
Will close this. Current discussion is in #434 and most of it has been already done it #429 .
